jina/math/lsh.py

```diff
@@ -0,0 +1,53 @@
+import numpy as np
+
+class LSH:
+    """
+    param: k: the number of hyperplanes
+    param: d: the dimensionality of the random hyperplanes 
+    """
+
+    def __init__(self, k, d):
+        self.k = k
+        self.d = d
+        self.H = np.random.normal(0, 1, (k, d))
+
+    def hash_to_binary_code(self, U, nr_perms, nr_coordinates):
+        """
+        param: U: data to be distributed to bins
+        param: nr_perms: how many permutations to use
+        param: nr_coordinates: the prefix length in each permutation
+        """
+        assert self.H.shape[1] == U.shape[1]
+        hashes = np.sign(self.H.dot(U.T))
+        hashes = hashes.T
+        print("hashes shape", hashes.shape)
+        bins = [{} for _ in range(nr_perms)]
```

Comment:
`defaultdict` is a better choice: https://docs.python.org/3/library/collections.html